File: scripts/solar/solar-common/util/lang/StringUtil.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["6"] = 1,["7"] = 1,["8"] = 1,["10"] = 1,["11"] = 8,["12"] = 9,["13"] = 10,["15"] = 12,["16"] = 8,["17"] = 19,["18"] = 20,["19"] = 19,["20"] = 28,["21"] = 28,["22"] = 28,["24"] = 29,["25"] = 30,["26"] = 30,["27"] = 30,["28"] = 30,["29"] = 31,["30"] = 28,["31"] = 36,["32"] = 37,["33"] = 38,["34"] = 36,["35"] = 42,["36"] = 43,["37"] = 43,["39"] = 44,["40"] = 45,["41"] = 46,["42"] = 47,["43"] = 48,["44"] = 49,["46"] = 51,["48"] = 42,["49"] = 56,["50"] = 57,["51"] = 58,["52"] = 59,["53"] = 60,["54"] = 61,["55"] = 62,["57"] = 64,["58"] = 56,["59"] = 68,["60"] = 69,["61"] = 70,["62"] = 71,["63"] = 68,["64"] = 94,["65"] = 95,["66"] = 94});
local ____exports = {}
____exports.default = __TS__Class()
local StringUtil = ____exports.default
StringUtil.name = "StringUtil"
function StringUtil.prototype.____constructor(self)
end
function StringUtil.isEmpty(self, str)
    if str == nil or str == "" then
        return true
    end
    return false
end
function StringUtil.notEmpty(self, str)
    return not ____exports.default:isEmpty(str)
end
function StringUtil.prototype.toFixed(self, num, fractionDigits)
    if fractionDigits == nil then
        fractionDigits = 2
    end
    local stringTs = string
    local lm = stringTs.format(
        ("%." .. tostring(fractionDigits)) .. "f",
        num
    )
    return lm
end
function StringUtil.getBytes(self, char)
    local code = string.byte(char)
    return ____exports.default:getBytesByCode(code)
end
function StringUtil.getBytesByCode(self, code)
    if not code then
        return 0
    end
    if code > 240 then
        return 4
    elseif code > 225 then
        return 3
    elseif code > 192 then
        return 2
    else
        return 1
    end
end
function StringUtil.getLength(self, str)
    local length = 0
    local currentIndex = 1
    while currentIndex <= #str do
        local code = string.byte(str, currentIndex)
        currentIndex = currentIndex + ____exports.default:getBytesByCode(code)
        length = length + 1
    end
    return length
end
function StringUtil.subString(self, str, start, ____end)
    local start_byte = utf8.offset(str, start + 1)
    local end_byte = utf8.offset(str, ____end + 1) - 1
    return string.sub(str, start_byte, end_byte)
end
function StringUtil.getCharAtIndex(self, str, index)
    return ____exports.default:subString(str, index, index)
end
return ____exports

File: scripts/solar/solar-common/util/lang/StringUtil.js
export default class StringUtil {
    /**
     * 是否是空的字符串
     * @param str
     */
    static isEmpty(str) {
        if (str == null || str == "") {
            return true;
        }
        return false;
    }
    /**
     * 是否 不是空的字符串
     * @param str
     */
    static notEmpty(str) {
        return !StringUtil.isEmpty(str);
    }
    /**
     * 返回一个以定点表示法表示数字的字符串。
     * @param num
     * @param fractionDigits
     */
    toFixed(num, fractionDigits = 2) {
        let stringTs = string;
        let lm = stringTs.format("%." + fractionDigits + "f", num);
        return lm;
    }
    /**获取字符的字节数*/
    static getBytes(char) {
        let code = string.byte(char);
        return StringUtil.getBytesByCode(code);
    }
    /**获取字符的字节数*/
    static getBytesByCode(code) {
        if (!code)
            return 0;
        if (code > 240) {
            return 4;
        }
        else if (code > 225) {
            return 3;
        }
        else if (code > 192) {
            return 2;
        }
        else {
            return 1;
        }
    }
    /**获取字符串的长度*/
    static getLength(str) {
        let length = 0;
        let currentIndex = 1;
        while (currentIndex <= str.length) {
            let code = string.byte(str, currentIndex);
            currentIndex = currentIndex + StringUtil.getBytesByCode(code);
            length++;
        }
        return length;
    }
    /**截取子字符串(索引从0开始) 支持utf8*/
    static subString(str, start, end) {
        let start_byte = utf8.offset(str, start + 1);
        let end_byte = utf8.offset(str, end + 1) - 1; // -- 注意这里是 end_char + 1 的偏移减一
        return string.sub(str, start_byte, end_byte);
        // let tempStr: string = str;
        // let byteStart: number = 1;
        // let byteEnd: number = -1;
        // let index: number = 0;
        // let bytes: number = 0;
        // end = end ?? str.length;
        // while (tempStr.length > 0) {
        //     if (index == start) {
        //         byteStart = bytes + 1;
        //     } else if (index == end) {
        //         byteEnd = bytes;
        //         break;
        //     }
        //     let code: number = string.byte(tempStr);
        //     bytes += StringUtil.getBytesByCode(code);
        //     tempStr = string.sub(str, bytes + 1);
        //     index++;
        // }
        // return string.sub(str, byteStart, byteEnd);
    }
    /**获取字符串(索引从0开始)*/
    static getCharAtIndex(str, index) {
        return StringUtil.subString(str, index, index);
    }
}
